File: Sem-2-LP2/Prac 2 (A_ Algorithm)/Graph/AStarGraph.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stdlib.h>
#include <limits>
#include <algorithm>
using namespace std;

struct Node {
    int vertex;
    int fScore;

    Node(int vertex, int fScore) : vertex(vertex), fScore(fScore) {}

    bool operator<(const Node& other) const {
        return fScore > other.fScore;
    }
};

struct Edge {
    int dest;
    int weight;

    Edge(int dest, int weight) : dest(dest), weight(weight) {}
};

class Graph {
private:
    int numVertices;
    vector<vector<Edge>> adjList;

public:
    Graph(int numVertices) : numVertices(numVertices), adjList(numVertices) {}

    void addEdge(int source, int destination, int weight) {
        adjList[source].emplace_back(destination, weight);
    }

    vector<int> aStarAlgorithm(int start, int goal, const vector<int>& heuristic) {
        vector<int> g(numVertices, numeric_limits<int>::max()); 
        vector<int> f(numVertices, numeric_limits<int>::max()); 
        vector<int> parent(numVertices, -1); 

        g[start] = 0;
        f[start] = heuristic[start];

        priority_queue<Node> openList;
        openList.emplace(start, f[start]);

        while (!openList.empty()) {
            Node current = openList.top();
            openList.pop();
            int currentVertex = current.vertex;

            if (currentVertex == goal) {
            	
                vector<int> path;
                int vertex = goal;
                while (vertex != -1) {
                    path.push_back(vertex);
                    vertex = parent[vertex];
                }
                reverse(path.begin(), path.end());
                
                int pathCost = g[goal];  
    			cout << "Path cost: " << pathCost << endl;
                return path;
            }
            
			
            for (const Edge& neighbor : adjList[currentVertex]) {
                int neighborVertex = neighbor.dest;
                int tentativeG = g[currentVertex] + neighbor.weight;

                if (tentativeG < g[neighborVertex]) {
                	
                    parent[neighborVertex] = currentVertex;
                    g[neighborVertex] = tentativeG;
                    f[neighborVertex] = g[neighborVertex] + heuristic[neighborVertex];
                    openList.emplace(neighborVertex, f[neighborVertex]);
                }
                
            }
            
        }
		
        return {}; 
    }
};

int main() {
	int n;
    cout << "Enter the size of the graph: ";
    cin >> n;
    Graph graph(n);
    int size;
    cout << "Enter the size of input: ";
    cin >> size;

    cout << "Enter edges of the graph:" << endl;
    for (int i = 0; i < size; i++) {
        int j, k, w;
        cout << "Enter the value of " << (i + 1) << " edge and its weight: ";
        cin >> j >> k >> w;
        if (j < n && k < n) {
            graph.addEdge(j, k, w);
        } else {
            cout << "Invalid Input" << endl;
        }
    }
    vector<int> heuristic(n);
    cout << "Enter the heuristic values for the vertices of the graph:" << endl;
    for (int i = 0; i < n; i++) {
        cout << "Enter " << i << " vertex's heuristic value: ";
        cin >> heuristic[i];
    }

    int startVertex, goalVertex;
    cout << "Enter the starting vertex of the graph: ";
    cin >> startVertex;
    cout << "Enter the ending vertex of the graph: ";
    cin >> goalVertex;

   	vector<int> path = graph.aStarAlgorithm(startVertex, goalVertex, heuristic);
	if (!path.empty()) {
    cout << "Optimal path found:";
		for (int vertex : path) {
			cout << " " << vertex;
		}
	}
	else {
    	cout << "Path not found!" << endl;
	}
    
   
return 0;
}

File: Sem-2-LP2/Prac 2 (A_ Algorithm)/Graph/AStarGraph.py
import heapq
from typing import List

class Node:
    def __init__(self, vertex: int, fScore: int):
        self.vertex = vertex
        self.fScore = fScore

    def __lt__(self, other):
        return self.fScore < other.fScore

class Edge:
    def __init__(self, dest: int, weight: int):
        self.dest = dest
        self.weight = weight

class Graph:
    def __init__(self, numVertices: int):
        self.numVertices = numVertices
        self.adjList = [[] for _ in range(numVertices)]

    def addEdge(self, source: int, destination: int, weight: int):
        self.adjList[source].append(Edge(destination, weight))

    def aStarAlgorithm(self, start: int, goal: int, heuristic: List[int]) -> List[int]:
        g = [float('inf')] * self.numVertices
        f = [float('inf')] * self.numVertices
        parent = [-1] * self.numVertices

        g[start] = 0
        f[start] = heuristic[start]

        openList = []
        heapq.heappush(openList, Node(start, f[start]))

        while openList:
            current = heapq.heappop(openList)
            currentVertex = current.vertex

            if currentVertex == goal:
                path = []
                vertex = goal
                while vertex != -1:
                    path.append(vertex)
                    vertex = parent[vertex]
                path.reverse()

                pathCost = g[goal]
                print("Path cost:", pathCost)
                return path

            for neighbor in self.adjList[currentVertex]:
                neighborVertex = neighbor.dest
                tentativeG = g[currentVertex] + neighbor.weight

                if tentativeG < g[neighborVertex]:
                    parent[neighborVertex] = currentVertex
                    g[neighborVertex] = tentativeG
                    f[neighborVertex] = g[neighborVertex] + heuristic[neighborVertex]
                    heapq.heappush(openList, Node(neighborVertex, f[neighborVertex]))

        return []

# Example usage
n = int(input("Enter the size of the graph: "))
graph = Graph(n)

size = int(input("Enter the size of input: "))
print("Enter edges of the graph:")
for i in range(size):
    print(f"Enter the value of {i+1} edge and its weight: ", end="")
    j, k, w = map(int, input().split())
    if j < n and k < n:
        graph.addEdge(j, k, w)
    else:
        print("Invalid Input")

heuristic = []
print("Enter the heuristic values for the vertices of the graph:")
for i in range(n):
    print(f"Enter {i} vertex's heuristic value: ", end="")
    h = int(input())
    heuristic.append(h)

startVertex = int(input("Enter the starting vertex of the graph: "))
goalVertex = int(input("Enter the ending vertex of the graph: "))

path = graph.aStarAlgorithm(startVertex, goalVertex, heuristic)
if path:
    print("Optimal path found:", end="")
    for vertex in path:
        print(" ", vertex, end="")
    print()
else:
    print("Path not found!")
